libs/ngx-nuts-and-bolts/src/lib/table-state/table-state.helper.ts

```diff
@@ -0,0 +1,153 @@
+import { ActivatedRoute, Router } from '@angular/router';
+import { isEqual } from 'lodash';
+import { Observable } from 'rxjs';
+import { distinctUntilChanged, map } from 'rxjs/operators';
+
+export type Sort = 'asc' | 'desc' | '';
+
+export interface IPageInfo {
+	pageIndex: number;
+	pageSize: number;
+	length?: number;
+}
+
+export interface ISortInfo {
+	sortingKey: string;
+	sortDirection: Sort;
+}
+
+export enum TableQueryParam {
+	PAGE_INDEX = 'pageIndex',
+	PAGE_SIZE = 'pageSize',
+	SORT_KEY = 'sortKey',
+	SORT_DIRECTION = 'sortDirection',
+	CUSTOM_FILTERS = 'filters',
+}
+
+export function createPaginationObservable(route: ActivatedRoute): Observable<IPageInfo> {
+	return route.queryParamMap.pipe(
+		map((paramMap) => {
+			const pageIndex = parseInt(paramMap.get(TableQueryParam.PAGE_INDEX) ?? '0', 10);
+			const pageSize = parseInt(paramMap.get(TableQueryParam.PAGE_SIZE) ?? '10', 10);
+
+			return {
+				pageIndex,
+				pageSize,
+			};
+		}),
+		distinctUntilChanged(
+			(pageEvent1, pageEvent2) =>
+				pageEvent1.pageIndex === pageEvent2.pageIndex && pageEvent1.pageSize === pageEvent2.pageSize
+		)
+	);
+}
+
+export function createSortingObservable(route: ActivatedRoute): Observable<ISortInfo> {
+	return route.queryParamMap.pipe(
+		map((paramMap) => {
+			const sortingKey = paramMap.get(TableQueryParam.SORT_KEY) ?? '';
+			const sortDirection = (paramMap.get(TableQueryParam.SORT_DIRECTION) ?? '') as Sort;
+
+			return {
+				sortingKey,
+				sortDirection,
+			};
+		}),
+		distinctUntilChanged(isEqual)
+	);
+}
+
+export function createCustomFiltersObservable<TFilterValue extends Record<string, unknown> | null = null>(
+	route: ActivatedRoute,
+	router: Router
+): Observable<TFilterValue | null> {
+	return route.queryParamMap.pipe(
+		map((queryParamMap) => {
+			let filters = null;
+
+			try {
+				if (queryParamMap.has(TableQueryParam.CUSTOM_FILTERS)) {
+					filters = JSON.parse(atob(queryParamMap.get(TableQueryParam.CUSTOM_FILTERS) || ''));
```

Comment:
Is saving the filters in base64 common practice?

Reply:
I suppose filter object can grow significantly in this case we also don't have to worry about if filter object has any special characters which could potentially invalidate url ? Also filter object could have nested objects, with this approach I think that we don't have to worry about many of this cases. I'm open to suggestions on how to better handle this.

Reply:
Encoding the filters object to base64 is the simplest option that also covers nested objects in the filter value. Otherwise we would somehow have to flatten it or place the stringified object in the URL and I think that would look a bit uglier than base64 string. It also ensures that no strange characters end up in the URL.